### ISSUE TITLE
are dimensions in conjugate Wishart case set up correctly?

### ISSUE THREAD

**Issue by perrydv** [nimble-dev/nimble#213]
@danielturek can you please look at the conjugate sampler generated in our conjugate Wishart test in test-mcmc?  There is an issue with dimensions of contribution_R, which appear to assume that thisNodeSize == d.  Is that correct?  That appears to be the source of a compiler warning, since symbolically one dimension of the size of contribution_R is then "thisNodeSize", which is then used in a size-check expression generated by compilation of the cholesky expression.  That size-check expression  triggers a compiler warning that thisNodeSize might not have been set, which makes sense because thisNodeSize is only set inside the for loop and the compiler is worried that the for loop might have 0 iterations.  Is the logic of the sizes going in to  contribution_R ok?

**Comment by perrydv** [comment]
I will add that when I had to systematically change how the conjugate samplers are generated to work with the newNodeFxns system it is quite possible I made a mistake in this 2D case, like mixing up d for thisNodeSize for example.  But I think you will need to check it out.

**Comment by perrydv** [comment]
More details:

Set up Wishart conjugacy mcmc test case (copied from test-mcmc):

```
trueCor <- matrix(c(1, .3, .7, .3, 1, -0.2, .7, -0.2, 1), 3)
covs <- c(3, 2, .5)

trueCov = diag(sqrt(covs)) %*% trueCor %*% diag(sqrt(covs))
Omega = solve(trueCov)

n = 20
R = diag(rep(1,3))
mu = 1:3
Y = mu + t(chol(trueCov)) %*% matrix(rnorm(3*n), ncol = n)
M = 3
data <- list(Y = t(Y), n = n, M = M, mu = mu, R = R)

code <- nimbleCode( {
  for(i in 1:n) {
    Y[i, 1:M] ~ dmnorm(mu[1:M], Omega[1:M,1:M]);
  }
  Omega[1:M,1:M] ~ dwish(R[1:M,1:M], 4);
})

newDf = 4 + n
newR = R + tcrossprod(Y- mu)
OmegaTrueMean = newDf * solve(newR)

wishRV <- array(0, c(M, M, 10000))
for(i in 1:10000) {
  z <- t(chol(solve(newR))) %*% matrix(rnorm(3*newDf), ncol = newDf)
  wishRV[ , , i] <- tcrossprod(z)
}
OmegaSimTrueSDs = apply(wishRV, c(1,2), sd)
```

Make model and samplers:
```
m <- nimbleModel(code, constants = data,inits = list(Omega = OmegaTrueMean))
mcmc <- configureMCMC(m)
mcmc$getSamplerDefinition('Omega')
```

Code for the Omega sampler includes
```
    contribution_R <- array(dim = c(d, d))
    contribution_df <- 0
    for (iDep in 1:N_dep_dmnorm) {
        thisNodeSize <- dep_dmnorm_nodeSizes[iDep]
        contribution_R <- contribution_R + asCol(dep_dmnorm_values[iDep, 
            1:thisNodeSize] - dep_dmnorm_mean[iDep, 1:thisNodeSize]) %*% 
            (asRow(dep_dmnorm_values[iDep, 1:thisNodeSize] - 
                dep_dmnorm_mean[iDep, 1:thisNodeSize]) %*% dep_dmnorm_coeff[iDep, 
                1:thisNodeSize, 1:d])
        contribution_df <- contribution_df + 1
    }
    newValue <- rwish_chol(1, cholesky = chol(prior_R + contribution_R), 
        df = prior_df + contribution_df, scale_param = 0)
```

I think the dimensions of the expression following contribution_R is thisNodeSize x d, which would assume that thisNodeSize = d since contribution_R was set up as dxd.

This has the effect of telling the nimble compiler that contribution_R should be thisNodeSize x d. The `chol(prior_R + contribution_R)` generates a size check that involves `thisNodeSize`, which the C++ compiler realizes might not have been set and (on Windows) issues a warning.

Should there be indices on `contribution_R` in this big expression inside the for loop?  And/or should that `1:d` be `1:thisNodeSize` (created by line 710 of MCMCconjugacy.R)? 

For the moment I am removing the generation of "thisNodeSize <- 0" into conjugate samplers.  That was a false fix.

**Comment by danielturek** [comment]
Unfortunately I cannot look at this in detail now, I wish I could.  But I might be able to shed a some light on the mystery.  I think the apparent redundancy between `d` and `thisNodeSize` comes from the most recent expansion of the conjugacy system to allow for "ragged" dmnorm-dmnorm conjugate relationships, where the target `dmnorm` node is dx1, and the dependent `dmnorm` nodes may be of different sizes.  Hence the different expression for `thisNodeSize`, which in this Wishart case I agree will always be `d`, but in other cases, (specifically, "ragged" dmnorm-dmnorm conjugacy) these could be different.  So it appears redundant in this case, but is actually necessary to keep @paciorek happy.  Sorry I haven't fully processed the problem or cause of compiler warning, which I can/will do, but cannot now.

Is this waiting for me, in terms of CRAN or v0.6 release?

**Comment by perrydv** [comment]
Yes, I agree it is from such ragged cases, but I’m not seeing if it extends correctly to a 2D case such as Wishart.

Unfortunately yes this is waiting on you.

Or we could decide to live with a Windows compiler warning for now.  What I’m not sure about is whether the compiler warning has in fact pointed us towards  a potential bug.

**Comment by paciorek** [comment]
I believe the code as is is ok in that there is no bug for how users would
almost always use it, but it doesn't deal with a (probably very) special
case. I'm attaching some math on the Wishart conjugacy pointing out that
current handling of 'coeff' is not quite right, but should work in cases
where coeff is simply a scalar.

Still perusing this some to think about the best way forward.

**Comment by paciorek** [comment]
Ok, so I think the thing we want to do for now is only allow this case:

  y_i ~ dmnorm( mu_i, a_i Q) where a_i is scalar.

(Otherwise we'd be taking the user at his/her word that   A_i * Q (as we
currently allow) is positive definite and I don't think this would allow us
to have the more comment case of a_i * Q where a_i is scalar.)

So I think we want to allow only the scalar case and change all the
thisNodeSize in the dimension '2' case to be 'd', except for the 'coeff'
pieces which should just be
dep_dmnorm_coeff[iDep])
instead of
dep_dmnorm_coeff[iDep,  1:thisNodeSize, 1:d])

I can try to work on this, but am a bit afraid of breaking something
(actually more afraid of making the code be convoluted to deal with the dim
'2' case), plus I'm not sure how to check that a_i is scalar.



chris

**Comment by perrydv** [comment]
I think I could make it always use `1:d` instead of `1:thisNodeSize` in the 2-dimensional case.

But should the coeff pieces be `dep_dmnorm_coeff[iDep,  1:d, 1:d]`?

They are populated here:

```
    I <- identityMatrix(d)
    model[[target]] <<- I
    calculate(model, calcNodesDeterm)
    for (iDep in 1:N_dep_dmnorm) {
        dep_dmnorm_offset[iDep, 1:d, 1:d] <- model$getParam(dep_dmnorm_nodeNames[iDep], 
            "prec")
    }
    model[[target]] <<- I * 2
    calculate(model, calcNodesDeterm)
    for (iDep in 1:N_dep_dmnorm) {
        dep_dmnorm_coeff[iDep, 1:d, 1:d] <- model$getParam(dep_dmnorm_nodeNames[iDep], 
            "prec") - dep_dmnorm_offset[iDep, 1:d, 1:d]
    }

```
(that is the code as now generated) so it appears to be the component-wise numerically determined slope of the `prec` parameter with respect to the Wishart-distributed variable, or something like that, but anyway it is not scalar.

**Comment by paciorek** [comment]
No, the coeff pieces should be dep_dmnorm_coeff[iDep]  (i.e. scalar), and
then we would change the '%*% coeff' to '* coeff' in the list that
specifies conjugacies.

So as long as we check that the coeff is scalar then we'd only need to do
one of those numerically-determined slope calculations.

**Comment by perrydv** [comment]
Ok I think we’re going to have to wait because I don’t see how dep_dmnorm_coeff could be populated from a precision matrix but come out as a scalar.

**Comment by danielturek** [comment]
I admittedly have not followed all details here, but here are some thoughts:

The conjugate sampler functions are all code-generated.  So allowing the 1:thisNodeSize in the dmnorm-dmnorm (possibly ragged) case, but only allowing the size of 1:d in the wishart-dmnorm case breaks the very standardized nature of these.  That would be unfortunate.  To double-check, for the dmnorm-dmnorm, we admit the case:

mu[1:2] ~ dmnorm(mu0[1:2], prec0[1:2, 1:2])
y[1:3] ~ dmnorm(A[1:3, 1:2] %*% mu[1:2], prec_y[1:3, 1:3])

but for the wishart-dmnorm case, we do *not* allow any such ragged cases.  But come to think of it, I think this is ok, since there's no possible such case as:

Omega[1:2,1:2] ~ dwish(R[1:2,1:2], 4)
Y[1:3] ~ dmnorm(mu0[1:3], A[...,...] %*% Omega[1:2,1:2])    ## no possible A matrix

I understand Chris raised another point, that we should require the A matrix in the above code to be a scalar.  Unfortunately, that would also break the mold of the current system, since it assumes that for any 1D node (e.g., Omega), the coefficient is assumed to be 2D, and all calculations are done in that manner.

One thing we could do is enforce that the "link" for the wishart-dmnorm be identity, which would not allow any coefficient (matrix or scalar) in front of the Omega.  That would further restrict what's considered "conjugate", but would cover the case @paciorek is concerned about.  I think this would also remove the related compiler warning, since there would no longer be any term called dep_dmnorm_coeff[...].

Thoughts on how to proceed?

**Comment by paciorek** [comment]
I"m fine with the idea of enforcing the linear link and omitting the coeff
entirely. That deals with the main use case for Wishart (not clear to me
that the multiplication by a scalar in the dependency is common) and we can
come back to how to make things more general when and if there is another
conjugacy for 2D random variables, which if exists is not common.

If we do this, will that remove the use of thisNodeSize in the Wishart
generated code? And if it is still in the 'values' object will that be an
issue?   I could look but will just ask because I am headed home...

**Comment by perrydv** [comment]
I think we can clear up some of these points more efficiently when we get back together to talk.

I have just made a pull request that, when the posterior is 2D (and is that only represented by Wishart?), always uses `1:d` instead of `1:thisNodeSize`.  This clears up the compiler warning I have on Windows.  Let's see if Travis-CI finds any problems and then run our full tests.

A few brief notes:

- Handling of "ragged" dmnorm-dmnorm conjugacy appears to assume that starting index is always 1.  i.e. it uses `1:thisNodeSize`, not something like `thisNodeStart:thisNodeEnd`.  
- For a 2D prior, it was already the case that ragged dependencies were not supported due to earlier code: see comments (that have been there a while) in the "2" clause of the `switch` statement in the `if(needsLinearityCheck)` part of `addPosteriorQuantitiesGenerationCode` in MCMC_conjugacy.R.  
- The Windows compiler warning was not spurious.  The fundamental problem was that the generated sampler contained code to add a d-by-d matrix to a thisNodeSize-by-d matrix.  NIMBLE's size processing to generate C++ code then (correctly) tracked the size expressions of an object as (thisNodeSize, d) and (correctly) generated a later run-time size check using those expressions.  One of those happened to occur in a place that triggered a Windows compiler warning that thisNodeSize might not have been set, and that was in fact a valid warning.
- The two-point slope determination is still used, so a scalar multiple of the Wishart-distributed variable, the case Chris emphasized, should still be handled correctly, I think.  I don't think matrix multiplication of the Wishart-distributed variable was ever close to being handled correctly (and not sure it would make sense to try).  I think there may have been some cases of confusion or miscommunication about what is represented by what variable in the generated code, which we can sort out talking in person.

**Comment by paciorek** [comment]
Sounds good. I'll work on tests today and hopefully submit to CRAN this
afternoon.

**Comment by paciorek** [comment]
It looks like matrix multiple of the Wishart-distributed variable is producing a conjugate updater even though that conjugacy is not handled correctly. I think we should see if we can turn off that conjucy detection in that case. Let's Daniel and I chat.

```
n = 2
R = diag(rep(1,3))
mu = 1:3
Y = mu + t(chol(trueCov)) %*% matrix(rnorm(3*n), ncol = n)
M = 3

code <- nimbleCode( {
  for(i in 1:n) {
      mymat[1:M,1:M] <- Sigma[1:M,1:M] %*% Omega[1:M,1:M] %*% Sigma[1:M,1:M]
    Y[i, 1:M] ~ dmnorm(mu[1:M], mymat[1:M,1:M])
  }
  Omega[1:M,1:M] ~ dwish(R[1:M,1:M], df = 4);
})

m <- nimbleModel(code, data = list(Y=t(Y)), constants = list(n=n, M=M,R=R),
                 inits = list(Sigma = diag(3)))

conf <- configureMCMC(m)
conf$getSamplers()
# conjugate_dwish_dmnorm sampler: Omega[1:3, 1:3]
```

**Comment by paciorek** [comment]
I've thought about this a bit more. In order for us to handle scalar multiples of the Wishart random variable in the precision of a dependency, we need to do a bunch of extra calculations and we have an incorrect result in the (probably unlikely) case of a user trying a matrix multiple of the Wishart random variable. Given that, I'm now inclined to just handle cases where the link is the identity and avoid conjugacy when that is not the case. 

This would mean that we would only detect conjugacy in a setting like this:
Y ~ dnorm(mu, W)
W ~ dwishart(R, df)

but not
Y ~ dnorm(mu, a*W)

If that's ok with Daniel and Perry, we'll simply change the link to identity and remove 'coeff' from the contribution_R calculation.

**Comment by danielturek** [comment]
See PR #276 <https://github.com/nimble-dev/nimble/pull/276>.

**Comment by perrydv** [comment]
@danielturek it appears from the final comment on this issue that you may have addressed it.  Can you confirm and if so close the issue?

**Comment by danielturek** [comment]
Yes, closed after merging #276

**Comment by paciorek** [comment]
As a followup note, one of my collaborators has a model she would like to fit that has a scalar multiplier; here the target node of interest is 'pf' and 'k_0' prevents us from recognizing conjugacy.  The comments in the issue here indicate the can of worms that opens, but wanted to note this for possible discussion. 

```
tobit2space.model <- nimbleCode({
  for(i in 1:N){
    y.censored[i,1:J] ~ dmnorm(muf[1:J], cov = pf[1:J,1:J])
    for(j in 1:J){
      y.ind[i,j] ~ dconstraint(y.censored[i,j] > 0)
    }
  }
  
  cov[1:J,1:J] <- k_0 * pf[1:J,1:J]
  muf[1:J] ~ dmnorm(mean = mu_0[1:J], cov = cov[1:J,1:J])
  
  Sigma[1:J,1:J] <- lambda_0[1:J,1:J]/nu_0
  pf[1:J,1:J] ~ dinvwish(S = Sigma[1:J,1:J], df = J)
})
```
